[master < T154-MAGE] Add k-means clustering docs (#570)
memgraph/docs: mage/algorithms/machine-learning-graph-analytics/kmeans-clustering-algorithm.md

---
id: kmeans-clustering-algorithm
title: K-Means Clustering
sidebar_label: K-Means Clustering
---

## Description

> The KMeans algorithm clusters given data by trying to separate samples in `n` groups of equal variance, by minimizing a criterion known as 

Reviewer: ```suggestion
The K-Means algorithm clusters given data by trying to separate samples in `n` groups of equal variance by minimizing a criterion known as 
```

> The k-means algorithm clusters given data by trying to separate samples in `n` groups of equal variance by minimizing a criterion known as 

Reviewer 2: ```suggestion
The k-means algorithm clusters given data by trying to separate samples in `n` groups of equal variance by minimizing the criterion known as 
```

within-the-cluster sum-of-squares. 

> Algorithm requires the number of clusters to be specified beforehand. This means it is not possible for algorithm to determine
> what is the best number of clusters to use, but it is up to user to find it. It scales well to large numbers of samples and has been used 
> across a large range of application areas in many different fields such as recommendation engine, fraud detection and so on.

Reviewer 2: Correcting grammar + redundancies
```suggestion
The algorithm requires the number of clusters to be specified beforehand. In other words, the algorithm doesn’t compute the optimal number of clusters to use, but leaves that task to the user. It scales well to large numbers of samples, and it has been applied in a wide range of fields such as recommendation engines, fraud detection and market research.
```

![k-means clustering](../../data/algorithms/machine-learning-graph-analytics/kmeans-clustering.png)

> > K-means clustering run on top of Node2Vec algorithm. Numbers represent cluster.

Reviewer 2: ```suggestion
> K-means clustering run on node2vec output; clusters are numbered.
```

Reviewer 2: BTW, is the node2vec part important here? I’ve looked at the above image and don’t see how it relates to node2vec

Author: It is not important, I will remove it

### Implementation

[![K-Means](https://img.shields.io/badge/KMeans-Implementation-FB6E00?logo=github&style=for-the-badge)](https://github.com/memgraph/mage/blob/main/python/kmeans.py)

[![K-Means](https://img.shields.io/badge/KMeans-Documentation-FCC624?style=for-the-badge&logo=python&logoColor=white)](/mage/query-modules/python/kmeans-clustering)

> **K-Means** algorithm is implemented within project

Reviewer: ```suggestion
**K-Means** algorithm is implemented within the
```

> [**MAGE**](https://github.com/memgraph/mage). Be sure to check it out in the

Reviewer: ```suggestion
[**MAGE**](https://github.com/memgraph/mage) project. Be sure to check it out in the
```

> **K-Means** algorithm is implemented within the
> [**MAGE**](https://github.com/memgraph/mage) project. Be sure to check it out in the
> link above.

Reviewer 2: ```suggestion
The **K-Means** algorithm is implemented within the
[**MAGE**](https://github.com/memgraph/mage) project. Be sure to check it out in the
link above.
```

## Use cases

[![Social
networks](https://img.shields.io/badge/Social_networks-Application-8A477F?style=for-the-badge)](/use-cases/social-media.md)

> Clustering is mostly used in social networks. Communities
> can be explored, similarities found, and much more.

Reviewer 2: ```suggestion
Clustering is heavily used in social network analysis to discover user communities, find similarities between them and much more.
```

> Biggest use case for clustering is in social networks. There communities
> can be explored, similarities found, and many more.

Reviewer: ```suggestion
Clustering is mostly used in social networks. Communities
can be explored, similarities found, and much more.
```

## Description

The k-means algorithm clusters given data by trying to separate samples in `n` groups of equal variance by minimizing the criterion known as 
within-the-cluster sum-of-squares. 
The algorithm requires the number of clusters to be specified beforehand. In other words, the algorithm doesn’t compute the optimal number of clusters to use, but leaves that task to the user. It scales well to large numbers of samples, and it has been applied in a wide range of fields such as recommendation engines, fraud detection and market research.

![k-means clustering](../../data/algorithms/machine-learning-graph-analytics/kmeans-clustering.png)

### Implementation

[![k-means](https://img.shields.io/badge/KMeans-Implementation-FB6E00?logo=github&style=for-the-badge)](https://github.com/memgraph/mage/blob/main/python/kmeans.py)

[![k-means](https://img.shields.io/badge/KMeans-Documentation-FCC624?style=for-the-badge&logo=python&logoColor=white)](/mage/query-modules/python/kmeans)

> **K-means** algorithm is implemented within the

Reviewer 2: ```suggestion
The **K-means** algorithm is implemented within the
```
